src/big/big_decimal.cr

```diff
@@ -0,0 +1,287 @@
+require "big_int"
+
+# A `BigDecimal` represents arbitrary precision decimals.
+#
+# It is internally represented by a pair of `BigInt` and `UInt64`: value and scale.
+# Value contains the actual value, and scale tells the decimal point place.
+# e.g. value=1234, scale=2 => 12.34
+#
+# The general idea and some of the arithmetic algorithms were adapted from
+# the MIT/APACHE -licensed https://github.com/akubera/bigdecimal-rs
+
+class InvalidBigDecimalException < Exception
+  def initialize(big_decimal_str : String, reason : String)
+    super("Invalid BigDecimal: #{big_decimal_str} (#{reason})")
+  end
+end
+
+struct BigDecimal
+  ZERO                       = BigInt.new(0)
+  TEN                        = BigInt.new(10)
+  DEFAULT_MAX_DIV_ITERATIONS = 100_u64
+
+  include Comparable(Number)
+  include Comparable(BigDecimal)
+
+  getter value : BigInt
+  getter scale : UInt64
+
+  # Creates a new `BigDecimal` from a `String`.
+  #
+  # Allows only valid number strings with an optional negative sign.
+  def initialize(str : String)
+    raise InvalidBigDecimalException.new(str, "Zero size") if str.bytesize == 0
+
+    # Check str's validity and find index of .
+    decimal_index = nil
+    str.each_char_with_index do |char, index|
+      case char
+      when '-'
+        if index != 0
+          raise InvalidBigDecimalException.new(str, "Unexpected '-' character")
+        end
+      when '.'
+        if decimal_index
+          raise InvalidBigDecimalException.new(str, "Unexpected '.' character")
+        end
+
+        decimal_index = index
+      when '0'..'9'
+        # Pass
+      else
+        raise InvalidBigDecimalException.new(str, "Unexpected #{char.inspect} character")
+      end
+    end
+
+    if decimal_index
+      value_str = String.build do |builder|
+        # We know this is ASCII, so we can slice by index
+        builder.write(str.to_slice[0, decimal_index])
+        builder.write(str.to_slice[decimal_index + 1, str.bytesize - decimal_index - 1])
+      end
+
+      @value = value_str.to_big_i
+      @scale = (str.bytesize - decimal_index - 1).to_u64
+    else
+      @value = str.to_big_i
+      @scale = 0_u64
+    end
+  end
+
+  # Creates an new `BigDecimal` from `Int`.
+  def initialize(num : Int)
+    initialize(num.to_big_i, 0)
+  end
+
+  # Creating a `BigDecimal` from `Float` is not supported due to precision loss risks. This call fails at compile time.
+  def initialize(num : Float)
+    {% raise "Initializing from Float is risky due to loss of precision -- use Int, String or BigRational" %}
```

Comment:
The only thing i'm still concerned about is this ctor. I don't like the idea of refusing to convert just because it's lossy. There are plenty of lossy transforms already in the stdlib, they should be documented, not unimplemented.

Reply:
I like the strictness here. However, initializing from a BigRational is lossy as well, isn't it? (Think 1/3, for example).

Reply:
I prefer idea of documenting losses and may be warnings about.

Reply:
I'm also drifting towards supporting these but warning/documenting. Would a compile-time warning be ok?

Reply:
@mverzilli true... and also, we don't even support initializations from BigRational :)

Reply:
I made a change where initialization and to_big_d from Float works, but outputs a warning to STDERR

Reply:
Please, no warnings. I can't imagine a program running with this code while STDERR gets flooded. I can also imagine the specs now output to STDERR?

Reply:
Ok, removed the warning. Devs will bear the responsiblity, then.

Reply:
@RX14 Does this satisfy your worries?